IdentityCore/src/cache/mac/MSIDMacTokenCache.m

```diff
@@ -22,39 +22,448 @@
                    context:(id<MSIDRequestContext>)context
                      error:(NSError **)error
 {
-    return nil;
+    MSID_LOG_INFO(context, @"itemWithKey:serializer:context:error:");
+    NSArray<MSIDToken *> *items = [self itemsWithKey:key serializer:serializer context:context error:error];
+    
+    if (items.count > 1)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorCacheMultipleUsers, @"The token cache store for this resource contains more than one user.", nil, nil, nil, context.correlationId, nil);
+        }
+        
+        return nil;
+    }
+    
+    return items.firstObject;
 }
 
-- (BOOL)removeItemWithKey:(MSIDTokenCacheKey *)key
-                  context:(id<MSIDRequestContext>)context
-                    error:(NSError **)error
+- (BOOL)removeItemsWithKey:(MSIDTokenCacheKey *)key
+                   context:(id<MSIDRequestContext>)context
+                     error:(NSError * __autoreleasing *)error
 {
-    return YES;
+    [self.delegate willWriteCache:self];
+    __block BOOL result = NO;
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        result = [self removeItemsWithKeyImpl:key context:context error:error];
+    });
+    [self.delegate didWriteCache:self];
+    
+    return result;
 }
 
 - (NSArray<MSIDToken *> *)itemsWithKey:(MSIDTokenCacheKey *)key
                             serializer:(id<MSIDTokenSerializer>)serializer
                                context:(id<MSIDRequestContext>)context
-                                 error:(NSError **)error
+                                 error:(NSError * __autoreleasing *)error
+{
+    [self.delegate willAccessCache:self];
+    NSArray *result = nil;
+    result = [self itemsWithKeyImpl:key serializer:serializer context:nil error:error];
+    [self.delegate didAccessCache:self];
+    
+    return result;
+}
+
+- (BOOL)saveWipeInfoWithContext:(id<MSIDRequestContext>)context
+                          error:(NSError **)error
+{
+    return NO;
+}
+
+- (NSDictionary *)wipeInfo:(id<MSIDRequestContext>)context
+                     error:(NSError **)error
 {
     return nil;
 }
 
+#pragma mark - Private
+
+- (void)addToItems:(nonnull NSMutableArray *)items
+    fromDictionary:(nonnull NSDictionary *)dictionary
+               key:(nonnull MSIDTokenCacheKey *)key
+{
+    MSIDToken *item = [dictionary objectForKey:[self keyWithoutAccount:key]];
+    if (item)
+    {
+        item = [item copy];
+        [items addObject:item];
+    }
+}
+
+- (void)addToItems:(nonnull NSMutableArray *)items
+         forUserId:(nonnull NSString *)userId
+            tokens:(nonnull NSDictionary *)tokens
+               key:(MSIDTokenCacheKey *)key
+{
+    NSDictionary *userTokens = [tokens objectForKey:userId];
+    if (!userTokens)
+    {
+        return;
+    }
+    
+    // Add items matching the key for this user
+    if (key.service)
+    {
+        [self addToItems:items fromDictionary:userTokens key:key];
```

Comment:
just curious, what is the case where key.service is not available?

Reply:
For example when you are querying all tokens for user:
```
MSIDTokenCacheKey *queryKey = [[MSIDTokenCacheKey alloc] initWithAccount:@"test_account" service:nil type:nil];
NSError *error;

NSArray<MSIDToken *> *items = [self.dataSource itemsWithKey:queryKey serializer:self.serializer context:nil error:&error];
```